File: Homework_Three_Four/conjgrad.m

function y = conjgrad(A, b, x)
    tolerance = 1.0e-5;
    denom = b' * b;
    % disp(denom)
    r = A * x - b;
    p = -r;
    rsold = r' * r;   % 34

    for i = 1:length(b)
        Ap = A * p;
        alpha = rsold / (p' * Ap);
        x = x + alpha * p;
        r = r + alpha * Ap;
        rsnew = r' * r;
        rnrm = sqrt(rsnew/denom)
        y = x;
        if rnrm < tolerance
              break;
        end
        p = -r + (rsnew / rsold) * p;
        rsold = rsnew;
    end
end

File: Homework_Three_Four/conjgrad.jl
function conjgrad(A,b,x)
    tolerance::Float64 = 1e-5
    denom::Float64 = only(b' * b)

    r = A*x - b

    p = -r

    rsold::Float64 = only(r' * r) # 34

    rnrm::Float64 = 1

    for i in 1:length(b)
        Ap = A*p

        alpha::Float64 = rsold / only(p' * Ap)

        x = x + alpha*p
 
        r = r + alpha*Ap

        rsnew::Float64 = only(r' * r )

        rnrm = sqrt(rsnew/denom)


        if rnrm < tolerance
            print("Took ",i," iterations\n\n")
            break
        end
        p = -r + (rsnew / rsold)*p

        rsold = rsnew

    end
    return x
end